File: lua/backlinks.lua
-- Backlinks manager for markdown files
-- Lua 5.1 compatible

local M = {}

function M.find_links(file_path)
    local file = io.open(file_path, 'r')
    if not file then return nil end
    
    local content = file:read('*all')
    file:close()
    
    local lines_before_bk = {}
    local lines_after_bk = {}
    local backlinks_found = false
    
    for line in content:gmatch('[^\n]+') do
        if not backlinks_found then
            table.insert(lines_before_bk, line)
            if line:match('^backlinks:%s*$') then
                backlinks_found = true
            end
        else
            table.insert(lines_after_bk, line)
        end
    end
    
    if not backlinks_found then
        print('In "' .. file_path .. '" not found "backlinks:" string')
        return nil
    end
    
    local links = {}
    for link in table.concat(lines_before_bk, '\n'):gmatch('%[([^%]]+)%]%(([^)]+)%)') do
        table.insert(links, link)
    end
    
    local backlinks = {}
    for line in table.concat(lines_after_bk, '\n') do
        for link in line:gmatch('%[([^%]]+)%]%(([^)]+)%)') do
            table.insert(backlinks, '[' .. link .. '](' .. link .. ')')
        end
    end
    
    return {links, backlinks}
end

function M.format_links(links)
    local result = {}
    for _, link in ipairs(links) do
        local filename = link:match('%((.+)%)')
        if filename then
            table.insert(result, '[' .. filename .. '](' .. filename .. ')')
        end
    end
    return result
end

function M.extract_filename(link)
    return link:match('%((.+)%)') or ''
end

function M.filename_to_markdown_link(file)
    local name = file:gsub('%.md$', ''):gsub('_', ' ')
    return '[' .. name .. '](' .. file .. ')'
end

function M.create_links_database(files)
    local database = {}
    for _, file_path in ipairs(files) do
        local links = M.find_links(file_path)
        if links then
            links[1] = M.format_links(links[1])
            local basename = file_path:match('([^/\\]+)$')
            database[basename] = links
        end
    end
    return database
end

function M.find_new_backlinks(database, verbose)
    local new_backlinks = {}
    
    for key, data in pairs(database) do
        for _, link in ipairs(data[1]) do
            local filename = M.extract_filename(link)
            if not database[filename] then
                goto continue
            end
            
            if not new_backlinks[filename] then
                new_backlinks[filename] = {}
            end
            
            local new_backlink = M.filename_to_markdown_link(key)
            local is_found = false
            
            for _, backlink in ipairs(database[filename][2]) do
                if link:find(backlink, 1, true) then
                    is_found = true
                    break
                end
            end
            
            if not is_found then
                for _, old_bk in ipairs(database[filename][2]) do
                    if new_backlink:find(old_bk, 1, true) then
                        is_found = true
                        break
                    end
                end
            end
            
            if not is_found then
                local found = false
                for _, bk in ipairs(new_backlinks[filename]) do
                    if bk == new_backlink then
                        found = true
                        break
                    end
                end
                if not found then
                    table.insert(new_backlinks[filename], new_backlink)
                    if verbose then
                        print('new backlink ' .. new_backlink .. ' in ' .. filename .. ' from ' .. key)
                    end
                end
            end
            
            ::continue::
        end
    end
    
    return new_backlinks
end

function M.add_new_backlinks(info)
    for file, bks in pairs(info) do
        local lines = {}
        local file_handle = io.open(file, 'r')
        if not file_handle then
            print('Cannot open file: ' .. file)
            goto next_file
        end
        
        for line in file_handle:lines() do
            table.insert(lines, line)
        end
        file_handle:close()
        
        local inserted = false
        for i, line in ipairs(lines) do
            if line:match('^backlinks:%s*$') then
                for _, bk in ipairs(bks) do
                    table.insert(lines, i + 1, '- ' .. bk)
                end
                inserted = true
                break
            end
        end
        
        if not inserted then
            print('In "' .. file .. '" not found "backlinks:" string')
            goto next_file
        end
        
        file_handle = io.open(file, 'w')
        if file_handle then
            for _, line in ipairs(lines) do
                file_handle:write(line .. '\n')
            end
            file_handle:close()
        end
        
        ::next_file::
    end
end

return M

File: modules/backlinks.py
import re
import os
from typing import List, Optional, Tuple, Dict, Any

LINKS = 0
BACKLINKS = 1


def read_lines_from_file(filename: str) -> List[str]:
    with open(filename, 'r', encoding='utf-8') as file:
        return file.readlines()


def write_lines_to_file(filename: str, lines: List[str]) -> None:
    with open(filename, 'w', encoding='utf-8') as file:
        file.writelines(lines)


def find_links(file_path: str) -> Optional[List[List[str]]]:
    """Find wiki links and backlinks in a markdown file.
    
    Args:
        file_path: Path to the markdown file
        
    Returns:
        List containing [links, backlinks] or None if backlinks section not found
    """
    backlinks = []
    links = []
    lines_before_backlinks = []
    lines_after_backlinks = []
    backlinks_found = False

    lines = read_lines_from_file(file_path)

    for line in lines:
        if not backlinks_found:
            lines_before_backlinks.append(line)
            if line.strip() == "backlinks:":
                backlinks_found = True
        else:
            lines_after_backlinks.append(line)

    if not backlinks_found:
        print(f"In \"{file_path}\" not found \"backlinks:\" string")
        return None

    regex = r"(\[.*\]\(.*\.md\))"
    matches = re.finditer(regex, "".join(lines_after_backlinks), re.MULTILINE)
    for match in matches:
        backlinks.append(match.group(1))

    regex = r"\[.*\]\(.*\.md\)"
    matches = re.finditer(regex, "".join(lines_before_backlinks), re.MULTILINE)
    for match in matches:
        links.append(match.group())

    return [links, backlinks]


def format_links(links: List[str]) -> List[str]:
    """Format markdown links to [filename](filename) format."""
    result = []
    for link in links:
        new_link = re.sub(r'\[.*\]\((.*?)\)', r'[\1](\1)', link)
        result.append(new_link)
    return result


def extract_filename(link: str) -> str:
    """Extract filename from markdown link."""
    return re.sub(r'\[.*\]\((.*?)\)', r'\1', link)


def filename_to_markdown_link(file: str) -> str:
    """Convert filename to markdown link format."""
    name = re.sub(r'(.*?).md', r'\1', file)
    name = name.replace("_", " ")
    return f"[{name}]({file})"


def create_links_database(files: List[str]) -> Dict[str, List[Any]]:
    """Create database of links from all markdown files.
    
    Args:
        files: List of file paths
        
    Returns:
        Dictionary with filename as key and [links, backlinks] as value
    """
    database = {}
    for file_path in files:
        links = find_links(file_path)
        if links is None:
            continue
        links[LINKS] = format_links(links[LINKS])
        database[os.path.basename(file_path)] = links
    return database


def find_new_backlinks(database: Dict[str, List[Any]], verbose: bool = False) -> Dict[str, List[str]]:
    """Find new backlinks that should be added to files.
    
    Args:
        database: Links database from create_links_database
        verbose: Enable verbose logging
        
    Returns:
        Dictionary with filename as key and list of new backlinks as value
    """
    new_backlinks = {}
    for key in database.keys():
        for link in database[key][LINKS]:
            filename = extract_filename(link)
            if filename not in database.keys():
                continue

            if filename not in new_backlinks.keys():
                new_backlinks[filename] = []

            new_backlink = filename_to_markdown_link(key)
            is_found = False
            
            for backlink in database[filename][BACKLINKS]:
                if link in backlink:
                    is_found = True

            if not is_found:
                for old_backlink in database[filename][BACKLINKS]:
                    if new_backlink in old_backlink:
                        is_found = True

            if not is_found and new_backlink not in new_backlinks[filename]:
                new_backlinks[filename].append(new_backlink)
                if verbose:
                    print(f"new backlink {new_backlink} in {filename} from {key}")

    return new_backlinks


def add_new_backlinks(info: Dict[str, List[str]]) -> None:
    """Add new backlinks to files.
    
    Args:
        info: Dictionary with filename as key and list of backlinks to add
    """
    for file in info.keys():
        lines = read_lines_from_file(file)
        backlinks_found = False
        for i, line in enumerate(lines):
            if line.strip() == "backlinks:":
                backlinks_found = True
                for bk in info[file]:
                    lines.insert(i + 1, f"- {bk}\n")
                break
        if not backlinks_found:
            print(f"In \"{file}\" not found \"backlinks:\" string")
            continue

        write_lines_to_file(file, lines)